### ISSUE TITLE
No way to pass PowerShell Arrays over the REST API?

### ISSUE THREAD

**Issue by haxtibal** [Icinga/icinga-powershell-apichecks#3]
For example, without the API, you would invoke a plugin like
``` 
> Invoke-IcingaCheckUsedPartitionSpace -Include 'C:','D:'
``` 
where `'C:','D:'` is somewhat magically transformed to a PowerShell Array object by the parameter binder.

Now with the REST API and without the `Exit-IcingaExecutePlugin` wrapper, but using curl or any http library, we would have to first serialize arguments to a JSON map like `{"Include": "'C:','D:'"}` and POST it as body to `/v1/checker?command=Invoke-IcingaCheckUsedPartitionSpace`.

This fails. The plugin function is actually called, but it outputs [UNKNOWN]. And I _believe_ it's because in `function Invoke-IcingaApiChecksRESTCall`, `Invoke-Command -ScriptBlock` propagates `'C:','D:'` as string without transforming it into an Array object. For verification, repeating the same with only `'C:'` as parameter works.

Can this be solved by serializing the JSON somehow differently? Or is it a bug in Invoke-IcingaApiChecksRESTCall.psm1?

Btw.: `Exit-IcingaExecutePlugin` doesn't work either for the described case, it reports an error and falls back to local plugin execution. I don't mind that much, my actual intention is to write a native alternative to `Exit-IcingaExecutePlugin` in Rust, to get rid of the powershell.exe initialization performance penalty. Will share it with you once ready.

**Comment by LordHepipud** [comment]
I did some tests in this matter and I'm not sure on where the problem occurs from. When using the Framework and check execution by enabling the REST-Api feature, this is an example on what is send over the API:

```json
"Status": "Running",
"Service": [
    "icinga2",
    "test$ding"
],
"Exclude": [ ],
"Verbosity": 3
}
```

This is the `Invoke-IcingaCheckService` plugin with `Service` being the array which includes two services for monitoring.

**Comment by LordHepipud** [comment]
Test call from Linux with `curl`:

```bash
curl --insecure https://icingahost:5668/v1/checker?command=Invoke-IcingaCheckService -X POST -d '{"Service": ["icinga2", "test"], "Verbosity": 3}' --header "Content-Type: application/json"
```

```json
{"Invoke-IcingaCheckService":{"exitcode":3,"checkresult":"[UNKNOWN] Services: 1 Unknown 1 Ok [UNKNOWN] test: Service not found (0) (All must be [OK])\n\\_ [OK] Service \"Icinga 2 (icinga2)\": Running\n\\_ [UNKNOWN] test: Service not found: 0","perfdata":["\u0027pending_paused_services\u0027=0;; ","\u0027running_services\u0027=0;; ","\u0027pending_continued_services\u0027=0;; ","\u0027stopped_services\u0027=0;; ","\u0027pending_started_services\u0027=0;; ","\u0027service_count\u0027=1;; ","\u0027pending_stopped_services\u0027=0;; ","\u0027paused_services\u0027=0;; ","\u0027test_service_not_found\u0027=;; ","\u0027service_icinga_2_icinga2\u0027=4;;4 "]}}
```

The encoding on curl is a bit rough, but for me it seems to work. Do I miss something?

**Comment by haxtibal** [comment]
Oh yes, I see. You're passing a readily prepared JSON array `"Service": ["icinga2", "test"]` in the request. In contrast, I was always passing strings like `"Service": "icinga2,test"` because I thought it's the REST API side who is responsible for interpreting special PowerShell expressions, if any.

That clears up things a bit. On the other hand, I'm not sure how to deal with that new situation. From your perspective, you can easily convert user strings to an JSON array at client side, because your client is written in PowerShell which has that special handling built in. My client is intentionally not written in PowerShell, which leaves me with two options
- tell the i4w_callapi user to be careful to not use any PowerShell expressions in their command definitions EDIT: very bad idea, that's way too limiting because many of the `Invoke-*`  arguments take values of type array, and none of that arguments could be expressed at CLI level then
- or, implement handling for some common cases (like PS arrays) in Rust

With my PR applied it would be possible to let the service side do the parameter expansion, ~~while passing json arrays should stay functional too (did no test yet)~~. That would be quite comfortable from the perspective of i4w_callapi, any chance to convince you? ;-)

EDIT:
> while passing json arrays should stay functional too (did no test yet)

After reconsidering, no, that would probably never work reliable. We have to strictly define at JSON level how types are interpreted by the REST API. Either a JSON string is guaranteed to be passed as raw string parameter to the check command (current code), or it is guaranteed to be interpreted by the parameter at REST API side (my PR). Both behavior at same time is not possible. And same goes for arrays and other types.

**Comment by haxtibal** [comment]
How about adding a second endpoint to switch between behavior?

"Structured API", as it exists now. At API service side, the cmdlet is invoked using splatting.
```
curl --insecure -X POST --header "Content-Type: application/json" \
  https://icingahost:5668/v1/checker?command=Invoke-IcingaCheckService \
  -d '{"Service": ["icinga2", "test"], "Verbosity": 3}' 
````

New "interpreted API", suitable for tools like i4w_callapi. The request content would simply be one string representing the whole command line (the arguments part) as you would type it in an interactive PS session. At API service side, the cmdlet is invoked using a string scriptblock as in my PR, to force invocation of the parameter binder.
```
curl --insecure -X POST --header "Content-Type: text/plain" \
  https://icingahost:5668/v1/interpretedChecker?command=Invoke-IcingaCheckService \
  -d "-Service icinga2,test -Verbosity 3"
````

I can quite understand if you're not happy with that suggestion, because everything already works in your setup and the interpreted API might be subject to security issues.

It's just that I'm foreseeing nasty code at my side when I have to try to resemble a PowerShell command line parser in Rust, which is the consequence if `v1/checker` stays as is and i4w_callapi users shall be able to use command definitions that are at least close to what Icinga Director generates.

**Comment by LordHepipud** [comment]
I honestly would not want to let the API decide on what datatype is parsed and always ensure JSON is properly submitted to the API. There are other aspects in current setups I know of, on which the JSON input is created and used already.

Is there any way for you to analyse the argument input? From PowerShell side and from the Icinga DSL (regardless of Icinga 2 native conf or Icinga Director), I'm always ensuring that custom vars defined as array are seperated by `,`.

In case there is an easy chance to detect this, you could simply "split" the string on the `,`, remove the first whitespace and create an array based on the input. As PowerShell and the parser does not care about single values, it wouldn't be a problem to ship single elements as plain string:

```bash
curl --insecure https://icingahost:5668/v1/checker?command=Invoke-IcingaCheckService -X POST -d '{"Service": "icinga2", "Verbosity": 3}' --header "Content-Type: application/json"
```

This does still work and only requires work for multiple values, which in case you are using the DSL from Icinga for Windows is `,` separated, while Icinga arrays are whitespace separated.

If there is a way to detect that these are multiple values, regardless on how they are written

```powershell
-Service 'icinga2', 'test'
```

```powershell
-Service 'icinga2' 'test'
```

it would resolve the issue.

In addition you could also implement your own logic for something like this:

```powershell
-Service 'icinga2' -Service 'test'
```

powershell does not allow multiple arguments of the same type, but maybe Rust does.

**Comment by haxtibal** [comment]
> Is there any way for you to analyse the argument input? From PowerShell side and from the Icinga DSL (regardless of Icinga 2 native conf or Icinga Director), I'm always ensuring that custom vars defined as array are seperated by `,`.

Sure, one can look for `,` and handle it as comma operator that outputs an array. But I'm afraid that's not the end of the story:
- Now that `-MyParam first,second` gets special handling, you would still want a way to pass a literal comma like `-MyParam "this is string, my friend"`. So I'd have to implement PS-style quoting (single or double quotes are currently not handled specially, they're received from argv as characters inside a string and are passed to the API as characters inside a string).
- Now that `-MyParam "a string"` gets special handling, you would still want a way to pass a literal quote, like `-MyParam "watch out for \""`, i.e. implement PS-style escaping.
- I'd have also to recognize `@()` which becomes empty array, and `False` which becomes bool. And what about nested types, like arrays of bools, or arrays of arrays.

In other words, this gets one into some deeper rabbit hole of inventing a PowerShell-like parser. Or let's say PowerShell deserializer, because we can limit the story to data types, but ignore whole statements.

A completely different approach I could think of was to solve it at the generator level. Instead of letting i4w_callapi eat PowerShell syntax, I could specify that each CLI argument has to be a JSON string. Then the deserializer is readily available. That would shift the heavy lifting into a new command definition generator that outputs arguments in JSON syntax instead of PS syntax. I'm not sure how hard that would be and how well it would integrate with Icinga Director. You already have such a generator, do you think it could be extended and reused that way?

**Comment by LordHepipud** [comment]
I see the problem here. I think point 3 is not really relevant, as nested arrays are not going to happen with plugins - this would be way to complicated for configuration and in addition, it is as far as I know not supported yet on the Icinga Director.

Re-Using or rebuilding the current configuration creator would be an entire mess. What you can basically do, is have a look on the plain configuration creator and modify the parts you require: https://github.com/Icinga/icinga-powershell-framework/blob/master/lib/core/tools/Get-IcingaCheckCommandConfig.psm1#L434

What it does will run through the generated hashtable for the command definitions (which is the source for the Director baskets) and modify them there.

If you simply want to create our down DSL to build the array as JSON array, you simply modify the current DSL for this:

```cpp
var arr = macro("$IcingaCheckService_Array_Exclude$");
if (len(arr) == 0) {
    return "@()";
}
return arr.map(
    x => if (typeof(x) == String) {
        var argLen = len(x);
        if (argLen != 0 && x.substr(0,1) == "'" && x.substr(argLen - 1, argLen) == "'") {
            x;
        } else {
            "'" + x + "'";
        }
    } else {
        x;
    }
).join(",");
```

**Comment by LordHepipud** [comment]
In my oppinion  the easiest way would be, to load the string char by char, analyze if it starts with `'` or `"` and then look until the string closes. If it does not close, it might be a simple string element. If the value contains a whitespace, without closing `'` or `"`, we can assume it is an whitespace array.

If there is a `,` without a whitespace element, it could be both - agreed.

However, the new Icinga Director baskets will always render arrays like this:

```powershell
-Service 'icinga2','test';
```

Basically a "rough" way to go with, could be something like this:

* In case a string contains `,`, split it on the `,` to have multiple values inside the array
* Now analyze each array elements and check if at least one of them starts and ends with `'`, it is an array
* If no element contains this, check if the values are numeric
* If not all of those values are numeric, is it not an array

This logic should be enough to cover most errors and assume the input. Agreed, if your string contains an `,`, it will be splitted that way. But we could then in addition lookup, if the first char of the array element is `'`. If so, we can have a look on the next array element. If it contains `'` at the end, both array elements belong together and must be joined with `,` again.

I do however not see these scenarios. The only one I could think if right now, would be for event log monitoring to filter for messages.

**Comment by haxtibal** [comment]
Thanks a lot for your explanations. We can close the issue now, as you explained well why it's not an issue of icinga-powershell-apichecks, and I'm left with two viable options:
- Implement an incomplete PS deserializer. I'll do so by means of a [custom serde deserializer](https://serde.rs/impl-deserializer.html). It's basically an outstanding exercise on my personal todo list anyway, which then turns into something useful.
- Modify the plain configuration creator, let Icinga DSL create JSON arguments.

@log1-c This will take some time. Guess it's not reasonable to continue testing before at least one of that options is ready.
